File: frontend/src/services/database.service.ts
import { apiClient } from "@/lib/api-client";
import type { DatabaseConnection, ConnectionListResponse } from "@/types/dataset.types";

export interface ConnectRequest {
  name: string;
  type: string;
  config: {
    host: string;
    port: number;
    database: string;
    username: string;
    password: string;
    ssl: boolean;
  };
}

export interface TestConnectionResponse {
  status: string;
  latency_ms: number;
  tables_count: number;
}

export const databaseService = {
  /** Test connection without saving */
  testConnection: async (body: { type: string; config: ConnectRequest["config"] }): Promise<TestConnectionResponse> => {
    const { data } = await apiClient.post<TestConnectionResponse>("/database/test", body);
    return data;
  },

  /** Save a new connection */
  connect: async (body: ConnectRequest): Promise<DatabaseConnection> => {
    const { data } = await apiClient.post<DatabaseConnection>("/database/connect", body);
    return data;
  },

  /** List all saved connections */
  listConnections: async (): Promise<ConnectionListResponse> => {
    const { data } = await apiClient.get<ConnectionListResponse>("/database/connections");
    return data;
  },

  /** Delete a connection */
  deleteConnection: async (connectionId: string): Promise<void> => {
    await apiClient.delete(`/database/connections/${connectionId}`);
  },
};

File: backend/app/services/database_service.py
"""
Database Service handling external database connections
"""
import json
import logging
import uuid
from sqlalchemy import create_engine, text
from sqlalchemy.ext.asyncio import create_async_engine, AsyncEngine
from app.utils.encryption import decrypt_data
from app.database.base import async_session_maker
from app.models.data_source import DataSource
from app.utils.exceptions import ConnectionTestFailedError

logger = logging.getLogger(__name__)


async def get_connection_engine(connection_id: str | uuid.UUID) -> AsyncEngine:
    """Decrypt DataSource configuration and return a SQLAlchemy AsyncEngine."""
    from app.database.base import async_session_maker
    async with async_session_maker() as db:
        result = await db.get(DataSource, uuid.UUID(str(connection_id)))
        if not result:
            raise ConnectionTestFailedError("Connection source not found")

        # Decrypt config
        decrypted_json = decrypt_data(result.encrypted_config)
        config = json.loads(decrypted_json)

        # Build url based on database type
        db_type = result.type
        host = config.get("host")
        port = config.get("port")
        username = config.get("username")
        password = config.get("password")
        database = config.get("database")

        if db_type == "postgresql":
            url = f"postgresql+asyncpg://{username}:{password}@{host}:{port}/{database}"
        elif db_type == "mysql":
            url = f"mysql+aiomysql://{username}:{password}@{host}:{port}/{database}"
        elif db_type == "sqlite":
            url = f"sqlite+aiosqlite:///{database}"
        elif db_type == "sqlserver":
            url = f"mssql+pyodbc://{username}:{password}@{host}:{port}/{database}?driver=ODBC+Driver+17+for+SQL+Server"
        else:
            raise ValueError(f"Unsupported connection type: {db_type}")

        return create_async_engine(url, pool_pre_ping=True)


class DatabaseService:
    """Service for handling testing and CRUD operations on database connections."""

    @staticmethod
    async def test_connection(config: dict, db_type: str) -> dict:
        """Test database connection without saving."""
        host = config.get("host")
        port = config.get("port")
        username = config.get("username")
        password = config.get("password")
        database = config.get("database")

        # Select async driver for testing
        if db_type == "postgresql":
            url = f"postgresql+asyncpg://{username}:{password}@{host}:{port}/{database}"
        elif db_type == "mysql":
            url = f"mysql+aiomysql://{username}:{password}@{host}:{port}/{database}"
        elif db_type == "sqlite":
            url = f"sqlite+aiosqlite:///{database}"
        elif db_type == "sqlserver":
            url = f"mssql+pyodbc://{username}:{password}@{host}:{port}/{database}?driver=ODBC+Driver+17+for+SQL+Server"
        else:
            raise ConnectionTestFailedError("Unsupported connection type")

        try:
            # We create an async engine, connect, and execute a simple query
            engine = create_async_engine(url, connect_args={"timeout": 5} if db_type == "postgresql" else {})
            async with engine.connect() as conn:
                await conn.execute(text("SELECT 1"))
            await engine.dispose()
            return {"status": "success", "tables_count": 0}
        except Exception as e:
            logger.error("Database connection test failed: %s", e)
            raise ConnectionTestFailedError(str(e))
